src/pages/workspaces/workspace/tools/WorkflowView.js

```diff
@@ -88,50 +103,63 @@ class WorkflowView extends Component {
       _.keys(await workspace.entities())
     )
 
-    const { invalidInputs, invalidOutputs, methodConfiguration: config } =
-      await workspace.methodConfig(workflowNamespace, workflowName).validate()
+    const validationResponse = await workspace.methodConfig(workflowNamespace, workflowName).validate()
+    const { methodConfiguration: config } = validationResponse
+    const ioDefinitions = await Rawls.methodConfigInputsOutputs(config)
 
-    const processIO = _.flow(
-      _.update('inputs', preprocessIOList),
-      _.update('outputs', preprocessIOList)
-    )
-
-    const processedIO = processIO(await Rawls.methodConfigInputsOutputs(config))
+    const inputsOutputs = this.createIOLists(validationResponse, ioDefinitions)
 
     this.setState({
-      inputsOutputs: processedIO,
-      invalid: this.createInvalidIOMap(invalidInputs, invalidOutputs, config, processedIO),
-      config, entityTypes
+      config, entityTypes, inputsOutputs, ioDefinitions,
+      anyInvalidInput: _.some('error', inputsOutputs.inputs),
+      anyInvalidOutput: _.some('error', inputsOutputs.outputs)
     })
   }
 
+  createIOLists(validationResponse, ioDefinitions = this.state.ioDefinitions) {
+    const { invalidInputs, invalidOutputs, methodConfiguration: config } = validationResponse
+
+    const invalid = {
+      inputs: invalidInputs,
+      outputs: invalidOutputs
+    }
+
+    const process = ioKey => _.map(({ name, inputType, outputType, optional }) => {
+      const value = config[ioKey][name]
+      const [task, variable] = _.takeRight(2, _.split('.', name))
+      const type = (inputType || outputType).match(/(.*?)\??$/)[1] // unify, and strip off trailing '?'
+      const error = Utils.cond(
+        [optional, () => undefined],
+        [!value, () => 'This attribute is required'],
+        () => invalid[ioKey][name]
+      )
+
+      return { name, task, variable, optional, value, type, error }
+    })
+
+    return {
+      inputs: process('inputs')(ioDefinitions.inputs),
+      outputs: process('outputs')(ioDefinitions.outputs)
+    }
```

Comment:
FP trivia: `_.update` can be handy in cases where you want to 'modify' the branches of a tree while keeping the trunk intact.

Reply:
Yeah, I couldn't figure out how to write something that wasn't way harder to read.